core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2348,14 +2348,20 @@ sealed trait ZIO[-R, +E, +A]
    * Executes this workflow when value of the specified `FiberRef` satisfies the
    * predicate.
    */
-  final def whenFiberRef[S](ref: => FiberRef[S])(f: S => Boolean)(implicit trace: Trace): ZIO[R, E, Option[A]] =
-    self.whenZIO(ref.get.map(f))
+  final def whenFiberRef[S](ref: => FiberRef[S])(f: S => Boolean)(implicit trace: Trace): ZIO[R, E, (S, Option[A])] =
```

Comment:
Ooh, nice improvement. 

Perfect, that way you get the `S` (which made the decision) no matter what! ❤️